pkg/provider/data_source_regions.go

```diff
@@ -0,0 +1,134 @@
+	resp.TypeName = req.ProviderTypeName + "_regions"
+}
+
+func (r *regionsDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Description: "Datasource ID.",
+				Computed:    true,
+			},
+			"regions": schema.ListNestedAttribute{
+				Description: "Region information.",
+				Computed:    true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"region_id": schema.StringAttribute{
+							Description: "Region ID of the region.",
+							Computed:    true,
+						},
+						"name": schema.StringAttribute{
+							Description: "Region name of the region.",
+							Computed:    true,
+						},
+						"status": schema.StringAttribute{
+							Description: "Region status of the region.",
+							Computed:    true,
+						},
+						"continent": schema.StringAttribute{
+							Description: "Continent that region belongs to.",
+							Computed:    true,
+						},
+					},
+				},
+			},
+
+			"cloud_provider": schema.StringAttribute{
+				Description: "Cloud provider to list the regions. For example, \"aws\" or \"azure\".",
+				Required:    true,
+			},
+			"project_id": schema.StringAttribute{
+				Description: "BigAnimal Project ID.",
+				Required:    true,
+				Validators: []validator.String{
+					ProjectIdValidator(),
+				},
+			},
+			"query": schema.StringAttribute{
+				Description: "Query to filter region list.",
+				Optional:    true,
+			},
+			"region_id": schema.StringAttribute{
+				Description: "Unique region ID. For example, \"germanywestcentral\" in the Azure cloud provider, \"eu-west-1\" in the AWS cloud provider.",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+type regionsDataSourceModel struct {
+	ID            *string          `tfsdk:"id"`
+	Regions       []*models.Region `tfsdk:"regions"`
+	CloudProvider *string          `tfsdk:"cloud_provider"`
+	ProjectId     *string          `tfsdk:"project_id"`
+	Query         types.String     `tfsdk:"query"`
+	RegionId      *string          `tfsdk:"region_id"`
+}
+
+func (r *regionsDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var cfg regionsDataSourceModel
+	diags := req.Config.Get(ctx, &cfg)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	regions := []*models.Region{}
+	if cfg.RegionId != nil {
+		region, err := r.client.Read(ctx, *cfg.ProjectId, *cfg.CloudProvider, *cfg.RegionId)
+		if err != nil {
+			resp.Diagnostics.Append(fromErr(err, "Error reading region by id: %v", cfg.RegionId)...)
+			return
+		}
+		regions = append(regions, region)
+
+	} else {
+		respRegions, err := r.client.List(ctx, *cfg.ProjectId, *cfg.CloudProvider, cfg.Query.ValueString())
+		if err != nil {
+			return
```

Comment:
There is something wrong in this part of the code. 
We call the `r.client.List` and read the response or check if there is any error. 
If there is any error, we're not adding it to `resp.Diagnostics`. 
If I'm not mistaken, we need to `res.Diagnostics.AddError` this error, and then return the method. 

Otherwise, when I have an BA API error, such as `no token` etc., current implementation misses that.